plugin/pkg/managedseed/validator/admission.go

```diff
@@ -0,0 +1,373 @@
+	// Find a primary DNS provider in the list of shoot DNS providers
+	primaryProvider := findPrimaryDNSProvider(shoot.Spec.DNS.Providers)
+	if primaryProvider == nil {
+		return nil, nil
+	}
+	if primaryProvider.Type == nil {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("primary DNS provider of shoot %s does not have a type", kutil.ObjectName(shoot)))
+	}
+	if *primaryProvider.Type == gardencore.DNSUnmanaged {
+		return nil, nil
+	}
+
+	// Initialize a reference to the primary DNS provider secret
+	var secretRef corev1.SecretReference
+	if primaryProvider.SecretName != nil {
+		secretRef.Name = *primaryProvider.SecretName
+		secretRef.Namespace = shoot.Namespace
+	} else {
+		secretBinding, err := v.coreClient.Core().SecretBindings(shoot.Namespace).Get(ctx, shoot.Spec.SecretBindingName, clientkubernetes.DefaultGetOptions())
+		if err != nil {
+			return nil, apierrors.NewInternalError(fmt.Errorf("could not get secret binding %s/%s: %v", shoot.Namespace, shoot.Spec.SecretBindingName, err))
+		}
+		secretRef = secretBinding.SecretRef
+	}
+
+	return &gardencore.SeedDNSProvider{
+		Type:      *primaryProvider.Type,
+		SecretRef: secretRef,
+		Domains:   primaryProvider.Domains,
+		Zones:     primaryProvider.Zones,
+	}, nil
+}
+
+func (v *ValidateManagedSeed) getSeedDNSProviderForDefaultDomain(ctx context.Context, shoot *gardencore.Shoot) (*gardencore.SeedDNSProvider, error) {
+	// Get all default domain secrets in the garden namespace
+	defaultDomainSecrets, err := v.kubeClient.CoreV1().Secrets(v1beta1constants.GardenNamespace).List(ctx, metav1.ListOptions{
+		LabelSelector: labels.SelectorFromValidatedSet(map[string]string{
+			v1beta1constants.GardenRole: common.GardenRoleDefaultDomain,
+		}).String(),
+	})
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("could not list default domain secrets in namespace %s: %v", v1beta1constants.GardenNamespace, err))
+	}
+
+	// Search for a default domain secret that matches the shoot domain
+	for _, secret := range defaultDomainSecrets.Items {
+		provider, domain, includeZones, excludeZones, err := common.GetDomainInfoFromAnnotations(secret.Annotations)
+		if err != nil {
+			return nil, apierrors.NewInternalError(fmt.Errorf("could not get domain info from domain secret annotations: %v", err))
+		}
+
+		if strings.HasSuffix(*shoot.Spec.DNS.Domain, domain) {
+			var zones *gardencore.DNSIncludeExclude
+			if includeZones != nil || excludeZones != nil {
+				zones = &gardencore.DNSIncludeExclude{
+					Include: includeZones,
+					Exclude: excludeZones,
+				}
+			}
+
+			return &gardencore.SeedDNSProvider{
+				Type: provider,
+				SecretRef: corev1.SecretReference{
+					Name:      secret.Name,
+					Namespace: secret.Namespace,
+				},
+				Domains: &gardencore.DNSIncludeExclude{
+					Include: []string{domain},
+				},
+				Zones: zones,
+			}, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func findPrimaryDNSProvider(providers []gardencore.DNSProvider) *gardencore.DNSProvider {
```

Comment:
Can you use `helper.FindPrimaryDNSProvider` instead?

Reply:
No, because `helper.FindPrimaryDNSProvider` uses a versioned resource (`gardencorev1beta1`) and I need an unversioned one (`gardencore`).

Reply:
Oops, I just found out that there is another `helper.FindPrimaryDNSProvider` that uses unversioned resources ... will use it of course.